### PR TITLE
Refactor signature functions and surrounding code to be more C++

### PR DESCRIPTION
**Pull request by ni4**
This PR continues changes, started in PR #1323, transferring code from pointer-based functions to C++ types.

### CODE REVIEW

**Comment by ni4** [comment]
macOS CI fails because of GitHub issue, causing `homebrew  update` to fail: https://github.com/Homebrew/homebrew-core/issues/61772

File: src/lib/types.h [hunk 1]
```diff
@@ -445,19 +699,24 @@ typedef struct pgp_revoke_t {
...
+    std::vector<uint8_t> z_algs{};
     // key server preferences (pgp_key_server_prefs_t)
-    uint8_t *ks_prefs;
-    size_t   ks_pref_count;
+    std::vector<uint8_t> ks_prefs{};
     // preferred key server
-    uint8_t *key_server;
+    std::string key_server{};
+
+    void set_symm_algs(const uint8_t algs[], size_t len);
```
Comment (rrrooommmaaa) [inline src/lib/types.h:712]: Why don't we add `set_xxx_algs` overloads that take `vector` as parameters? It would simplify other parts of code.

Comment (ni4) [reply src/lib/types.h:712]: @rrrooommmaaa Hm, actually we don't require uint8_t algs[] here at all. Will update the PR.

**Review by rrrooommmaaa — APPROVED** [review]
LGTM!

**Comment by ronaldtse** [comment]
Re-running macOS tests since they failed.

**Review by antonsviridenko — APPROVED** [review]
LGTM,
but I think next time it would be nice to split such PRs into multiple smaller ones for easier review,
this one is too big

**Comment by ni4** [comment]
> but I think next time it would be nice to split such PRs into multiple smaller ones for easier review,
> this one is too big

Yeah, this one is huge a bit, so I tried to separate it to commits as much as possible, so it could be reviewed on per-commit basis.

**Comment by ni4** [comment]
@ronaldtse Should we wait few more days until homebrew/GitHub fixes itself so macOS test passes?

**Comment by ronaldtse** [comment]
Testing brew fix in #1335. Once that works, can you help rebase here? Thanks.

**Comment by ni4** [comment]
> Testing brew fix in #1335. Once that works, can you help rebase and merge here? Thanks.

Brew fix worked, thanks! Rebased (together with #1334), let's wait till CI finish here.

**Comment by ronaldtse** [comment]
Single failure here:

```sh
Failures:

  1) #<UnboundMethod: Rnp#symmetric_encrypt> with a single password raises an error with a bad password
     Failure/Error:
       expect do
         rnp.decrypt(input: Rnp::Input.from_string(encrypted))
       end.to raise_error(Rnp::BadPasswordError)

       expected Rnp::BadPasswordError, got #<Rnp::Error: rnp_decrypt failed - (rc: 0x11000001): Error reading file> with backtrace:
         # ./lib/rnp/error.rb:35:in `raise_error'
         # ./lib/rnp/utils.rb:21:in `call_ffi'
         # ./lib/rnp/rnp.rb:546:in `block in decrypt'
         # ./lib/rnp/output.rb:138:in `default'
         # ./lib/rnp/rnp.rb:545:in `decrypt'
         # ./spec/encrypt/simple_encrypt_spec.rb:150:in `block (4 levels) in <top (required)>'
         # ./spec/encrypt/simple_encrypt_spec.rb:149:in `block (3 levels) in <top (required)>'
     # ./spec/encrypt/simple_encrypt_spec.rb:149:in `block (3 levels) in <top (required)>'

Finished in 1 minute 14.84 seconds (files took 0.57033 seconds to load)
239 examples, 1 failure

Failed examples:

rspec ./spec/encrypt/simple_encrypt_spec.rb:146 # #<UnboundMethod: Rnp#symmetric_encrypt> with a single password raises an error with a bad password
```

**Comment by ni4** [comment]
> Single failure here:

@ronaldtse Hm, it seems to be all fine now. Could we merge it?

**Comment by ni4** [comment]
Ping @ronaldtse

**Comment by ronaldtse** [comment]
Merged. Thanks!

**Pull request merged**